src/batMap.ml

```diff
@@ -171,6 +171,36 @@ module Concrete = struct
       | Empty -> raise Not_found in
     loop map
 
+  let find_first_opt f map =
+    let rec loop found tree =
+      match tree with 
+      | Node(l, k, v, r, _) ->
+         if f k
+         then loop (Some (k, v)) l
+         else loop found r
+      | Empty -> found in
+    loop None map
```

Comment:
If we wanted to avoid the `Some (k, v)` in the middle, we could have two founctions, `loop_not_found f tree` and `loop_found f k v tree`. (I'm including `f` as a parameter to make the functions closed, which is interesting here.)

Reply:
Done. The loop_found and loop_not_found functions are nested in the find_first and find_last functions. If you want i can extract them, but I think that should not be necessary.